shell/browser/api/electron_api_web_contents.cc

```diff
@@ -2572,6 +2619,43 @@ std::vector<content::NavigationEntry*> WebContents::GetHistory() const {
   return history;
 }
 
+void WebContents::RestoreHistory(
+    v8::Isolate* isolate,
+    gin_helper::ErrorThrower thrower,
+    int index,
+    const std::vector<v8::Local<v8::Value>>& entries) {
+  if (!web_contents()
+           ->GetController()
+           .GetLastCommittedEntry()
+           ->IsInitialEntry()) {
+    thrower.ThrowError(
+        "Cannot restore history on webContents that have previously loaded "
+        "a page.");
+    return;
+  }
+
+  auto navigation_entries = std::make_unique<
+      std::vector<std::unique_ptr<content::NavigationEntry>>>();
+
+  for (const auto& entry : entries) {
+    content::NavigationEntry* nav_entry = nullptr;
+    if (!gin::Converter<content::NavigationEntry*>::FromV8(isolate, entry,
+                                                           &nav_entry) ||
+        !nav_entry) {
+      // Invalid entry, bail out early
+      return;
+    }
+    navigation_entries->push_back(
+        std::unique_ptr<content::NavigationEntry>(nav_entry));
+  }
+
+  if (!navigation_entries->empty()) {
+    web_contents()->GetController().Restore(
+        index, content::RestoreType::kRestored, navigation_entries.get());
```

Comment:
Adding a restore API was previously attempted in https://github.com/electron/electron/pull/42014/commits/e3830e6a1fe49ea92903cabde39ed67fb98a0e16

Is there any part of that implementation which would be worth carrying over here? `entry->SetIsOverridingUserAgent(true)` looked like something to test. When these entries are loaded, does it restore the old user agent?

Reply:
We don't support `is_overriding_user_agent` in our loading code, so there's no point in supporting it here, either.

The API discussed there was a little different as it let you modify history on a "live" webContents without actually using pageState.